Add CliCredential

wantedfast opened the pull request:
Follow PR :  https://github.com/Azure/azure-sdk-for-python/pull/8962/file

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-sdk-for-net?pullRequest=9106) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-sdk-for-net?pullRequest=9106) before we can accept your contribution.<br/><br/>:x: wantedfast [sign now](https://cla.opensource.microsoft.com/Azure/azure-sdk-for-net?pullRequest=9106)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-sdk-for-net?pullRequest=9106) it.</sub>

ellismg commented on sdk/identity/Azure.Identity/src/CliCredentialClient.cs:
```diff
@@ -0,0 +1,65 @@
+            }
+            else
+            {
+                fileName = "/bin/sh";
```
On *nix, is it not possible to just invoke `az` directly?  The script itself is marked executable and has the correct shebang so I believe that we should be able to invoke it directly?  Are we unable to do this because we don't know the full path to the executable?

wantedfast replied:
The command can’t execute if there don’t have the full path on *nix.

ellismg replied:
Gotcha.

ellismg replied:
Gotcha

ellismg commented on sdk/identity/Azure.Identity/src/CliCredentialClient.cs:
```diff
@@ -0,0 +1,65 @@
+            string fileName = string.Empty;
+            string argument = string.Empty;
+
+            Process proc = new Process();
```
`Process` is disposable, we should have a `using` guarding this.

ellismg commented on sdk/identity/Azure.Identity/src/CliCredentialClient.cs:
```diff
@@ -0,0 +1,65 @@
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                fileName = "cmd";
+                argument = $"/c \"{extendCommand}\"";
```
Do we need to escape shell-meta characters in any way from `extendedCommand` since it did not expect to invoked via `cmd`?

wantedfast replied:
This extendCommand is a combination of  "get account get-access-token --resource" and "Scopes". Can we double check with a person who knows "Scopes" clear?

ellismg replied:
My concern is that someone could call `GetToken` directly, passing whatever scopes they want.  If that scope contains shell meta characters, it's possible that we could force `cmd` to run additional programs after the Azure CLI.  I don't know if this is something we need to be worried about.

ellismg replied:
Looking at https://github.com/Azure/go-autorest/blob/b965d219bd9d10676384ef8f05da2e2441df9829/autorest/azure/cli/token.go#L125-L175 may be useful.  They do some validation of "resource" there (which for us would be the "scope" that's put into the extended command.  I think if we did that validation we'd be in good shape.

wantedfast commented on sdk/identity/Azure.Identity/src/CliCredential.cs:
```diff
@@ -0,0 +1,133 @@
+
+                if (exitCode != 0)
+                {
+                    bool isLoginError = output.StartsWith("Please run 'az login'", StringComparison.CurrentCultureIgnoreCase);
```
I tested in many environment with many languages like Chinese and Japanese, all of them return English error messages. 
Looks like Azure CLI don’t provide localized error message. Let me know if it is not this case.

ellismg replied:
Do you know if `cmd` localizes it's error message?  I would wonder if on a Chinese and Japanese machine if `WinAzureCLIError`, which is string printed by `cmd` would be localized.  I guess no mater what we'll end up throwing an exception and passing back a message that has the output of `cmd` so we are probably fine as is.

wantedfast commented on sdk/identity/Azure.Identity/src/CliCredential.cs:
```diff
@@ -0,0 +1,133 @@
+                Dictionary<string, string> result = await JsonSerializer.DeserializeAsync<Dictionary<string, string>>(stream, null, cancellationToken);
+                result.TryGetValue("accessToken", out string accessToken);
+                result.TryGetValue("expiresOn", out string expiresOnValue);
+                DateTimeOffset expiresOn = DateTimeOffset.Parse(expiresOnValue, null, DateTimeStyles.AdjustToUniversal);
```
I cannot find an official time format of az cli tool according Azure CLI docs,.
Currently, expiresOn format is "yyyy-MM-dd HH:mm:ss.ffffff". So ParseExact that format. What do you think ?

ellismg replied:
Sounds good!

wantedfast commented on sdk/identity/Azure.Identity/src/CliCredential.cs:
```diff
@@ -0,0 +1,133 @@
+                Dictionary<string, string> result = await JsonSerializer.DeserializeAsync<Dictionary<string, string>>(stream, null, cancellationToken);
+                result.TryGetValue("accessToken", out string accessToken);
+                result.TryGetValue("expiresOn", out string expiresOnValue);
+                DateTimeOffset expiresOn = DateTimeOffset.Parse(expiresOnValue, null, DateTimeStyles.AdjustToUniversal);
```
This cli command(get account get-access-token) running in Process. And I tried, it never return string N/A.
May be we don’t need to consider N/A here. And we don’t need manual set the ExpiresOn to vary far.

ellismg replied:
> May be we don’t need to consider N/A here.

I do believe that if we ended up running this code from a program operating within Cloud Shell, we would see the N/A.  That may be out of scope for now and throwing is fine.  I suspect that when running in cloud shell usually, you'll get your credentials from somewhere else (like MS).

ellismg commented on sdk/identity/Azure.Identity/src/CliCredential.cs:
```diff
@@ -0,0 +1,135 @@
+            try
+            {
+                string command = ScopeUtilities.ScopesToResource(requestContext.Scopes);
+                string extendCommand = $"az account get-access-token --resource {command}";
```
We should explicitly pass `--output json` here, in case the user has a different default set.

wantedfast commented on sdk/identity/Azure.Identity/src/CliCredential.cs:
```diff
@@ -0,0 +1,133 @@
+
+                    if (isWinError || isOtherOsError)
+                    {
+                        throw new AuthenticationFailedException(AzureCLINotInstalled);
```
@ellismg Can we resolve this?

ellismg replied:
Yeah - I think what we have here is good.

jongio commented on sdk/identity/Azure.Identity/src/AzureCliCredential.cs:
```diff
@@ -0,0 +1,142 @@
+                }
+
+                byte[] byteArrary = Encoding.ASCII.GetBytes(output);
+                MemoryStream stream = new MemoryStream(byteArrary);
```
Shouldn't we be calling dispose (or use using statement) on any class that implements IDisposable?

wantedfast replied:
Detail in this [PR](https://github.com/wantedfast/azure-sdk-for-net/pull/1)

jongio commented on sdk/identity/Azure.Identity/src/AzureCliCredential.cs:
```diff
@@ -0,0 +1,142 @@
+    /// </summary>
+    public class AzureCliCredential : TokenCredential, IExtendedTokenCredential
+    {
+        private const string AzureCLINotInstalled = "Azure CLI not installed";
```
Be consistent with usage of CLI and Cli for .NET please use Cli

schaabs commented on sdk/identity/Azure.Identity/src/AzureCliCredentialClient.cs:
```diff
@@ -0,0 +1,68 @@
+                proc.Start();
+                proc.BeginOutputReadLine();
+                proc.BeginErrorReadLine();
+                proc.WaitForExit();
```
Should this wait have a timeout?

wantedfast replied:
Detail in this [PR](https://github.com/wantedfast/azure-sdk-for-net/pull/1)

schaabs commented on sdk/identity/Azure.Identity/src/AzureCliCredentialClient.cs:
```diff
@@ -0,0 +1,68 @@
+{
+    internal class AzureCliCredentialClient
+    {
+        public virtual (string, int) GetAzureCliAccesToken(string resource)
```
This method should probably return the actual access token so that all the process execution and output parsing is handled in one place, rather than split between this class and `AzureCliCredential`

wantedfast replied:
Detail in this [PR](https://github.com/wantedfast/azure-sdk-for-net/pull/1)

schaabs commented on sdk/identity/Azure.Identity/src/AzureCliCredential.cs:
```diff
@@ -0,0 +1,142 @@
+
+                    if (isWinError || isOtherOsError)
+                    {
+                        throw new AuthenticationFailedException(AzureCLINotInstalled);
```
Sorry I missed the question by @ellismg earlier. I agree this should be a `CredentialUnavailableException`,

schaabs replied:
Also the `IExtendedTokenCredential` interface is meant to provide an abstraction where exceptions are returned rather than thrown. This allows the caller to determin if the error should be raised.

wantedfast replied:
Detail in this [PR](https://github.com/wantedfast/azure-sdk-for-net/pull/1)

schaabs commented on sdk/identity/Azure.Identity/src/AzureCliCredentialClient.cs:
```diff
@@ -0,0 +1,68 @@
+                {
+                    FileName = fileName,
+                    Arguments = argument,
+                    UseShellExecute = false,
```
Should we be using shell execute here rather than starting the shell (cmd or /bin/sh)? This way we could be using exceptions to determine if the az command is not available rather than parsing the output of the shell process.

schaabs commented on sdk/identity/Azure.Identity/src/AzureCliCredential.cs:
```diff
@@ -0,0 +1,142 @@
+                    }
+                    else if (isLoginError)
+                    {
+                        throw new AuthenticationFailedException(AzNotLogIn);
```
This should also be a `CredentialUnavailableException`.

jsquire committed:
[Event Hubs Client] Test Infrastructure Retry Tweaks (Tracks One and Two) (#9766)

The focus of these changes is to tweak the conditions for retries when
creation of dynamic test infrastructure fails, based on observations of the
failure details.   This pass also does some light code clean-up, in an
attempt to improve quality and readability.

jsquire committed:
[Event Hubs Client] Mocked Async Event Source Workaround (#9774)

The focus of these changes is to work around an issue that appears to be
related to the test runner's asynchronous-to-synchronous bridge where an
IAsyncEnumerable<T> is used in conjunction with a background task.

A specific error is triggered within the CLR that I've not been able to
pin down to an absolute root cause.  I have been able to reproduce using
NUnit, a background task, and a variety of IAsyncEnumerable implementations
outside of the Event Hubs ecosystem.

Ultimately, the exception is benign but does potentially interfere with
proper testing of the exception handler.  This workaround addresses the
specific scenario.

maririos committed:
Renaming Batch methods (#9762)

* batch vs single

* documentation

* PR feedback

heaths committed:
Update Key Vault track 1 packages descriptions (#9781)

* Direct KV track 1 customers to track 2 packages

Resolves #9553 by updating the descriptions customers would see in UI package managers like in Visual Studio. Will also appear in the description on nuget.org, though after the current description (of which an exerpt shows in search results).

* Update Framework references assemblies

* Run analyze step on Windows

The Export-API.ps1 script was failing on non-Windows due to Framework
compatibility issues during build.

* Restore config to build track 2 instead of track 1

JoshLove-msft committed:
Don't run ApiCompat for design time build (#9777)

filizt committed:
[Deployment Scripts] Update .NET SDK version (#9788)

* Update SDK version

* Address offline review feedback

azure-sdk committed:
Sync eng/common directory with azure-sdk-tools repository (#9791)

maririos committed:
Create PiiEntity (#9790)

* piiEntity

* rename to entities

maririos committed:
Changes from UX study for Text Analytics  (#9772)

* uri -> url

* no DetectedLanguages

* add requestOptions to convenience batch methods

* bad delete

chidozieononiwu committed:
Produce and Publish Dependency Graph (#9784)

weshaggard committed:
Add * as text auto in gitattributes (#9512)

seanmcc-msft committed:
Added ShareFileClient.ClearRangeAsync() (#9684)

jsquire committed:
[Event Hubs Client] Marking Flaky Tests for Skip (Track One) (#9799)

The focus of these changes is to mark some tests that have recently begun
to display intermittent failures in CI or nightly runs until they can be
investigated and addressed.

pakrym committed:
Sanitize header values in exceptions (#9782)

chidozieononiwu committed:
Add InterdependencyGraph.html (#9805)

Grayer123 committed:
Regenerate 2019-10-01 SDK for Tag API (#9803)

* TagApi-2019-10-01 changes: Update Generated code; add tag related test cases; Update SessionRecords due to api version change.

* Fix Deployment Whatif test cases.

* Fix deployment test cases.

* Check in generate.ps1 releated changes, modified assembly version and release notes as well.

* Update AzSdk.RP.props.

* Regenerate models and operations from swagger 2019-10-01 for description improvement.

* Update tag related scenario tests as well as remove models no longer needed.

* Update assembly version and release notes.

maririos committed:
Rename TextSentiment to SentenceSentiment (#9801)

* sentenceSentiment

* rename TextSentimentLabel

pakrym committed:
Remove commit hash from telemetry header (#9813)

chidozieononiwu committed:
Add Script for adding Include paths to mgmt yml (#9567)

* Add Script for adding include paths to mgmt yml

* Set trigger to none, and adjust update-mgmt-yml script

* Update Contributing Docs

ellismg committed:
Revert "Support `AZURE_AUTH_LOCATION` (#9312)" (#9829)

This reverts commit 324fdf9b1eec15137823b0a41908b0d30526b5d9.

There is ongoing discussion about the best way to surface this support,
either in Azure.Identity itself, as a standalone sample or some other
way.

scbedd committed:
Resolve wonky spacing causing markdown header to not render as header in html (#9830)

Commit:
Azure.Identity simplify exception messaging (#9821)

* Azure.Identity simplify exception messaging

maririos committed:
Error model Text Analytics (#9817)

* error model

* PR feedback

maririos committed:
fix readme (#9840)

azure-sdk committed:
Sync eng/common directory with azure-sdk-tools repository (#9837)

JoshLove-msft committed:
Build sample apps (#9808)

nonik0 committed:
Updating to ADAL 5.2, VS token provider bug fix (#9797)

seanmcc-msft committed:
Data Lake API review changes (#9850)

pakrym committed:
Fix nullable warnings that appear only when compiled agains net… (#9858)

maririos committed:
Changelog for TA P2 (#9859)

* changelog

* typos

* mre typos

* update version

pakrym committed:
Update CHANGELOG.md for Azure.Core (#9861)

pakrym committed:
Update Azure.Core to 1.0.2 (#9870)

pakrym committed:
Update Azure.Core READMES (#9871)

Commit:
Updating Azure.Identity changelog and version for 1.1.1 release (#9883)

* Updating Azure.Identity changelog and version for 1.1.1 release

* PR feedback updates

areddish committed:
Areddish/training 3.1 (#7631)

* Regenerate for 3.1

* Add tests for new methods

* Update Session Records

* Update ids.txt

* Update id.txt path

* Update metadata to master branch

* Update version

seanmcc-msft committed:
Feature/storage/stg71 base (#9476)

heaths committed:
Add security policy (#9890)

maririos committed:
Update readme and docstrings (#9887)

* readme

* chu

* options

Co-Authored-By: Shawn Fang <45607042+mssfang@users.noreply.github.com>

* PR feedback

* format

Co-authored-by: Shawn Fang <45607042+mssfang@users.noreply.github.com>

christothes committed:
Extract Load Balancing from the Event Processor (#9198)

* changes for discussions

* tweaks to error handling

* project organization and PartitionLoadBalancerTests

* converted EventProcessorClientTests

* flakey test, log lastError

solankisamir committed:
Added support for DeveloperPortal and Open Api Json format (#9844)

* generated basic client

* record tests and changelog

* submit mgmt file

azure-sdk committed:
Increment package version after release of Azure.Identity (#9889)

maririos committed:
Fix sample links in text analytics (#9894)

* fix sample links

* Pr feedback

seanmcc-msft committed:
Prepare Storage for release (#9888)

azure-sdk committed:
Increment package version after release of Azure.AI.TextAnalytics (#9906)

pakrym committed:
Add Track 2 code generation support (#9885)

seanmcc-msft committed:
Updated storage release notes (#9910)

chidozieononiwu committed:
Update version bump tool (#9893)

* Update Version Bump tool to allow manual version overide

* Update Azure.Core Version

* Update script, update versioning.md, revert core version changes

Commit:
Updating Azure.Identity package references to use the latest release (#9903)

pakrym committed:
Add OPTIONS and TRACE HTTP request methods. (#9924)

AlexanderSher committed:
Remove failing tests for filtering that is not supported anymore. (#9929)

serkantkaraca committed:
Processor host changes to fix a race and better init performance (#9357)

* A race fix and init improvements in processor host

* Couple more improvements to avoid excessive task spin for large partition use cases.

* Spin up less number of tasks while adding and removing pumps.

serkantkaraca committed:
Fix breaking change (#9103)

* Moving SDK version to 4.1.0 for release

* Fixing breaking change at EventHubsException

Commit:
Azure.Identity removing IExtendedTokenCredential internal interface (#9932)

* Azure.Identity removing IExtendedTokenCredential internal interface

* Updating DefaultAzureCredential to only walk chain once

* addressing PR feedback

* adding test for selected credential caching

chidozieononiwu committed:
Filter out External dependencies before dumping data for graph (#9925)

maririos committed:
Add more tests to TA (#9849)

* add tests and one subcategory

* recordings

* new recording

* error countryhint

* check error details

* add id to string

JoshLove-msft committed:
Initial Track 2 Service Bus (#9941)

serkantkaraca committed:
ManagedIdentityTokenProvider to support custom AzureServiceTokenProvider (#9943)

* ManagedIdentityTokenProvider to support custom AzureServiceTokenProvider

* XmlDoc change

pakrym committed:
Add KeyVault.Secrets DataProtection extensions (#9953)

Marusyk committed:
Rename PartitionManager class to StorageManager (#9912)

* Rename PartitionManager class to StorageManager

jsquire committed:
[Service Bus Client] Initial Live Test Infrastructure (#9952)

The focus of these changes is to introduce infrastructure to allow the
Service Bus live tests to dynamically manage resources, creating a namespace
for each test run and allowing each test to request a dedicated queue or topic
scope.

Tests using this infrastructure are safe to run in parallel and isolated such
that tests are using individual Azure resources and will not cascade failures
nor should one test be able to impact the operation of another.

pakrym committed:
Add KeyVault.Secrets configuration extensions (#9937)

seanmcc-msft committed:
Fixed issue where Storage generated code and swagger didn't align (#9946)

chidozieononiwu committed:
Move Dependency Generation to Agregate Pipeline (#9964)

amnguye committed:
CreateIfNotExists/DeleteIfExists logs exception (#9551)

* CreateIfNotExists/DeleteIfExists for blobs logs exception and updated pipeline log method messages

serkantkaraca committed:
Fix blob call timeouts (#9988)

* Set ServerTimeout and MaximumExecutionTime for blob I/O requests.

* Instrument more blob APIs with request options.

jsquire committed:
[Service Bus Client] Test Resources Tweaks (#10009)

The focus of these changes is to correct an issue with the test resources
ARM template selecting an incorrect subscription for output, ensure the
service principal has the contributor role, to allow make some small
adjustments to template parameters.

maririos committed:
Improve naming in TA (#10022)

* DataSourceEntityId

* sentiment renaming

seanmcc-msft committed:
Added Exists, CreateIfNotExists, and DeleteIfExists to Data Lake (#9996)

JoshLove-msft committed:
Misc fixes (#10000)

seanmcc-msft committed:
Fixed Azure.Storage.Blobs Read-only file system exception bug (#10001)

serkantkaraca committed:
Timeout ListBlobsSegmentedAsync call until Storage issue is figured out (#10019)

* Set ServerTimeout and MaximumExecutionTime for blob I/O requests.

* Instrument more blob APIs with request options.

* Implement a workaround for ListBlobsSegmentedAsync hung issue.

* Throw OperationCancelledException instead.

* Cancel timer if ListBlobsSegmentedAsync returns successfuly.

ShivangiReja committed:
Implement receive batch methods (#10033)

 Implement Receive method

anuchandy committed:
Adding placeholder to list any third party licenses (#9825)

* Adding placeholder to list any third party licenses

* Adding standard header

pakrym committed:
Add Storage DataProtection integration package (#9961)

kinelski committed:
[EventHubs] Improving Event Hubs unit tests coverage (#9761)

* Improved Event Hubs unit tests coverage

* Removed duplicated live tests

heaths committed:
Use base64 encoding when importing certificates (#9998)

* Use base64 encoding for cer, csr, and x5c

Resolves #9963

* Reset challenge cache before each test

Resolves #9356

* Fix MergeCertificate operation

Fixes #9986 by passing the model to the request method.
Fixes #7893 by adding tests for merge (tests for import were added in a previous commit)

* Update changelog for previous commit

maririos committed:
Add none country hint (#10032)

* add none country hint

* add changelog

* Typo

* to const

azure-sdk committed:
Increment version for storage releases (#9914)

* Increment package version after release of Azure.Storage.Common

* Increment package version after release of Azure.Storage.Blobs

* Increment package version after release of Azure.Storage.Blobs.Batch

* Increment package version after release of Azure.Storage.Files.Shares

* Increment package version after release of Azure.Storage.Files.DataLake

* Increment package version after release of Azure.Storage.Queues

weshaggard committed:
dependencies -> dependency

maririos committed:
Separate document and sentence sentiment label (#10039)

* create sentence sentiment label

* remove warning

chidozieononiwu committed:
Enable Track One EventHubs Packages (#10049)

chidozieononiwu committed:
Fix safenames for eventhub ci (#10052)

chidozieononiwu committed:
Rename aggregate-reports pipeline tio match python and java repos (#10051)

seanmcc-msft committed:
Added DataLakeFileClient.Upload() (#9681)

pakrym committed:
Adjust keyvault extensions package names (#10034)

bitsforthought committed:
[Azure Cognitive Search] Fix broken codegen (#10054)

pakrym committed:
Move extensions projects to sdk/extensions directory (#10053)

azure-sdk committed:
Sync eng/common directory with azure-sdk-tools repository (#10048)

tg-msft committed:
Disable failing Storage tests (#10045)

This is being tracked by #10044.

maririos committed:
change cs to md (#10024)

kinelski committed:
Ignore tests failing during nightly runs (#10068)

JoshLove-msft committed:
Receive event (#10047)

yoavfr committed:
Azure Security Center - Update AdvancedThreatProtection (#9867)

* Update AdvancedThreatProtection

* updates

* udpdate test recordings

* run generatep.ps1 again

seanmcc-msft committed:
Improved the XML comments for creating Files and Directories in Data … (#10077)

jongio committed:
Update MI msg. (#9958)

weshaggard committed:
Removing broken link to blobs

jimmyca15 committed:
Fixed spelling error in ManagedIdentityCredential constructor summary. (#9012)

jongio committed:
Fix a couple of spelling/grammar issues. (#9141)

Commit:
Enabling client sdk analyzers for Azure.Identity (#10095)

* Enabling client sdk analyzers for Azure.Identity

* removing EnableClientSdkAnalyzers tag from csproj

kinelski committed:
[EventHubs] Ignore unstable test (Track Two) (#10098)

* Ignore unstable test

* One more

JoshLove-msft committed:
Await renew locks (#10100)

chidozieononiwu committed:
Update MacOS VM Images (#10070)

ShivangiReja committed:
[Service Bus] Implement settlement methods for received messages (#10104)

Implement message complete method

seanmcc-msft committed:
Added QueueClient.CreateIfNotExists(), .Exists(), add .DeleteIfExists() (#10050)

AlexanderSher committed:
Prepare source code for the sync/async mixed code analyzer (part 1) (#10065)

* Prepare source code for the sync/async mixed code analyzer (part 1)

* Missing files

* Address some of the CR comments

* Make TaskExtensions NUnit-friendly.

* Disable individual cases of AZC01XX rules violation instead of global rule disabling

* Add Debugger.Break() call

* Enable ClientSdkAnalyzers for Azure.Identity

* Fix release build

* Update package version

* Disable AZC0102 for Azure.Extensions

* Throw exception in VerifyTaskCompleted when debugger is attached

jsquire committed:
[Event Hubs Client] Add Event Source Traits (#10105)

The focus of these changes is to add the Azure SDK traits to each of the
event sources used by Event Hubs.

jsquire committed:
[Event Hubs Client] Test Resources Clean-Up (#10102)

The focus of these changes is to perform some clean-up and tweaks to the
live tests and their associated Infrastructure.  Included are:

  - Refactoring of the `TestEnvironment` and `EventHubScope` to simplify
    some of the logic and reduce levels of indirection.

  - Remove the test-specific `ConnectionStringTokenParser`, given that
    the standard `ConnectionStringParser` was required to be part of the
    shared code assets and can now be accessed by tests or explicitly Included
    in the test project.

  - Ensure that the dynamic resource creation infrastructure is ignoring the
    synchonization context for asynchronous operations.

  - Include a test resource creation template to integrate with the Engineering
    infrastructure to dynamically create/remove resource groups for test runs.

  - Refactoring the Consumer live tests to make use of common infrastructure for
    the "read until empty" loop that many were using.

  - Ensured that any `TaskCompletionSource` instances created are specifying
    `TaskCreationOptions.RunContinuationsAsynchronously` to avoid potential deadlocks.

Jingshu923 committed:
Db2 add connectionString and sftp add UseTempFileRename  (#10122)

* Update DB2 and Sftp

* add sdk info

fore5fire committed:
Fix case of env name Dogfood in Azure-Test-Framework doc (#9452)

aim-for-better committed:
[Synapse Control Plane]ON BOARD: add management sdk for new service Synapse (#10093)

* on board new management sdk for new service synapse

* Update SDK version with preview suffix

seanmcc-msft committed:
Added Exists(), CreateIfNotExists(), and DeleteIfExists() to ShareCli… (#10074)

jsquire committed:
[Event Hubs Client] Clean Nightly Test Configuration (#10116)

With the Event Hubs nightly tests making use of the test resources
template, the previous approach of setting environment variables from the
pipeline test configuration is no longer needed.

The focus of these changes is to remove them.

jsquire committed:
[Event Hubs Client] Documentation Updates (#10069)

The focus of these changes is to rename and enhance the migration guide and
to add the Event Processor proposal as a design asset.  Sample links have also
been made absolute to work around a tooling error.

jsquire committed:
[Event Hubs Client] Enable AppCompat Checks (#10127)

The focus of these changes is to enable the AppCompat tool to ensure that
API exposed by the the Event Hubs packages does not cause breaking changes
with the versions that have been GA'd.

Riding along are some minor formatting tweaks.

christothes committed:
Delete obsolete tests and fix some flaky tests (#10117)

* delete obsolete tests

* fixes some flaky tests for #7458

* delete commented line

* pr comments

JoshLove-msft committed:
Processor client (#10106)

rileymckenna committed:
reset to Azure/master (#9654)

wantedfast committed:
Change filed AzureCLINotInstalled, WinAzureCLIError to AzureCliNotInstalled,WinAzureCliError. And use shell.

jerryc3 committed:
add new api version for SIG (#10112)

emgu-ms committed:
Add PublicNetworkAccess + PrivateEndpointConnections (Get only) to Server (#10134)

* Failover database and elastic pools Net SDK

* Fixed tests and bumped up sdk version

* Regenerated using Azure/Master

* Copied sql_resource-manager.txt to the right place

* Add PublicNetworkAccess for set/get Server and PrivateEndpointConnections for get Server

* Fix .csproj

* Fix .csproj again

rickysun93 committed:
[Data Factory]Add properties into ADF SDK (#10140)

* Add additionalColumns/dataConsistency/sessionLog/isolationLevel/expiryDateTime/fileListPath/apiVersion to ADF SDK

* Add queryTrigger API and rerun from failure support

amolagar5 committed:
Move Server Azure AD Admin from 2014-04-01 to 2019-06-01-preview (#9959)

* sync to latest.

* Move Serer Azure Active Directory from 2014-04-01 to 2019-06-01-preview api.

* update SQL project version for updating Server Azure AD Admin API from 2014-04-01 to 2019-06-01-preview

* sync to latest

* Revert last checkin

* Move Server Azure AD from 2014-04-01 to 2019-06-01-preview

* add variables to record session.

* update version information.

* revert unrelated changes.

* reverted to master for appconfiguration

* Revert changes to PS-VSPrompt.lnk

Co-authored-by: Amol Agarwal <amagarwa@microsoft.com>
Co-authored-by: Yeming Liu <felix_liu@outlook.com>

christothes committed:
get some time back from CI (#10138)

pakrym committed:
Add response read timeouts (#10128)

JoshLove-msft committed:
Message model (#10154)

pakrym committed:
Remove dependency on Identity from extensions packages (#10146)

blueww committed:
[Storage] Support allowProtectedAppendWrites (#10141)

joshfree committed:
Add Gitter community channel (#9297)

Add Gitter community channel for azure-sdk-for-net

Commit:
Fix SDK Automation Failure (#10159)

jsquire committed:
[Event Hubs Client] Preparation for Event Processor<T> Implementation (#10147)

The focus of these changes is to set the groundwork for implementation of
the Event Processor base class by renaming members, shifting namespaces,
and flipping some internal types to public.

pakrym committed:
Revert "Add response read timeouts (#10128)" (#10168)

This reverts commit c4c10aca779f9d94aebf01efd5fb31d4bf25d7ea.

oleksandr-bilyk committed:
ServiceBus ManagedIdentityTokenProvider with custom  AzureServiceTokenProvider. (#9824)

pakrym committed:
Add response read timeouts (#10169)

aim-for-better committed:
[Synapse DataPlane]ON BOARD: Synapse DataPlane Sdk Initial Release (#10115)

* Synapse DataPlane Sdk Initial Release

* Add README.md to fix validate error

* Update version to 0.1.0-preview

* 1. Remove unnecessary file
2. Update README.md to add example
3. update csproj file

* Add CI config file

nemakam committed:
Updating ApiApproval for MSI (#10174)

Fixing test that started failing after #9824 was checked-in

AlexanderSher committed:
Fix #10109: Add version of EnsureCompleted for IAsyncEnumerable<T> (#10161)

* - Fix #10109: Add version of EnsureCompleted for IAsyncEnumerable<T>
- Add comments to the disabled diagnostics

* Fix comments

* Build fix

* Build fix

* Revert removal of using

* Added comments

yangyuan committed:
[Cognitive Services] Change SDK to stable (#10152)

filizt committed:
[Blueprints] Update .NET SDK: add excludedActions property; update documentation (#10066)

* Update SDK with swagger changes

* Add generated .txt file

* Update version

* Address reviewer notes

serkantkaraca committed:
Don't block PartitionManager loop on close pump call. (#10188)

* Don't block partition pump loop on close pump call.

* Comment fix

* Comment only change

kinelski committed:
[EventHubs] Distributed Tracing: Move message spans outside of Send spans (Track Two) (#10171)

kinelski committed:
[EventHubs] Distributed Tracing: Do not populate schema in peer.address attribute (Track Two) (#10183)

weshaggard committed:
Add test leg for testing against live packages for live test runs (#10170)

pakrym committed:
Add extensions live tests (#10195)

chidozieononiwu committed:
Extract release notes for use in Nuget packages (#10062)

* Extract release notes for use in nuget packages

* More work related to using MSBuild to extract ReleaseNotes

* Test release for Azure Template Preview 7

* Testing release with Azure.Template Previrew.6

* Testing release with Azure.Template Previrew.6

* Move all releasenotes logic into Extract-ReleaseNotes Script and capture Console output rather than writing temp file

maririos committed:
subcategories to string (#10192)

seanmcc-msft committed:
Fixed regression in BlobClient.Upload() and .DownloadTo() (#10099)

gkostal committed:
Integrated with latest swagger spec for management of policy certificates (#10172)

* Integrated with latest swagger spec for management of trusted policy signing certificates

* Updated backslash to forwardslash for referencing test data file

tg-msft committed:
Azure.Search preview (#10204)

Initial commit of the Azure.Search Track 2 client library for Azure Cognitive Search.

kinelski committed:
[EventHubs] Fixed EventBatch tests and updated batch constructor (Track Two) (#10218)

* Fixed EventBatch tests and included namespace + EHname to batch constructor

ellismg committed:
[eventhubs] Clarify MaximumWaitTime documentation (#10220)

When using the `EventProcessorClient` with a `MaximumWaitTime`, if the
timeout is reached, the `ProcessEventAsync` event is raised with an
empty `ProcessEventArgs` structure. The documentation refered to this as
a "null event" which is a little confusing, since `ProcessEventArgs`
itself is not null. Instead, the `HasEvent` property returns `false`.
Try to clarify this is a bit and add some remarks and cross references.

ShivangiReja committed:
[Service Bus] Implement message RenewLock and RenewSessionLock (#10223)

ramoka178 committed:
Tags on subscription and deployment (#9804)

* Changes corresponding to swagger changes

* clean up using correct namespace

* revert changes

* revert

* revert copywrite statement

* recorded new tests

* re-recorded the tests for deployment and subscription apis with tags info

* update version number

* update AssemblyFileVersion

* Re record tests after swagger is updated

* update Live tags test session records

gkostal committed:
Updated management SDK for Azure Attestation Service to latest swagger spec with Location and Tags support (#10212)

* Updated to latest swagger spec with Location and Tags support

* Fixed mix up in version number

* Delete code-model-v1

Co-authored-by: Yeming Liu <felix_liu@outlook.com>

christothes committed:
Speed up ReadEventsAsyncPublishesEventsWithxIterators tests (#10230)

* Speed up ReadEventsAsyncPublishesEventsWithxIterators tests

* add ConfigureAwait(false)

ellismg committed:
[identity] Support AZURE_CLIENT_CERTIFICATE_PATH (#10196)

This change adds support for setting `AZURE_CLIENT_CERTIFICATE_PATH` to
the path to an PFX or PEM. The private key corresponding to the
certificate must be unencrypted.

As part of this change, we now allow constructing a
`ClientCertificateCredential` from a path to a file on disk, in either
PFX or PEM format.

Supporting PFX is pretty straightforward, since there's great support
in the .NET. For PEM files, things are harder. While newer versions of
.NET Core have the primitives needed to be able to decode a PEM file and
get an `X509Certificate2` file, earlier versions of .NET Core (and
versions of .NET Framework) have gaps in support.

There are two major steps for reading these PEM files:

1. Decoding the PEM file to extract both the certificate and the private
key.
2. Attaching the private key to the `X509Certificate2` object.

For (1), The hard part is decoding the PKCS#8 encoded Private Key. In
.NET Core 3.0+, there is built in support via
`RSA.ImportPkcs8PrivateKey`, and on platforms which don't support this,
we've included a lightweight decoder based on some code from the .NET
Team that is able to decode this data. While the reader does not support
the entirity of the PKCS#8 spec, in practice it is sufficent for
decoding PEM files generated by common tools, for example, the Azure CLI
as a part of `az ad sp create-for-rbac --create-cert`.

For (2), There is an API which is present in .NET 4.7.2+ and .NET Core
2.0+, which allows you to assign the `RSA` object obtained by (1) to a
`X509Certificate2`.  While we can't depend on this method directly,
because we target NET Standard 2.0, we can do a reflection invoke.

Taken together, this allows us to support PEM certificates on .NET Core
2.0+ and .NET Framework 4.7.2.

From an implementation point of view, the constructors that take a path
to a PEM file do not actually attempt to load the file when the object
is being constructed (since this will require I/O) so instead we defer
reading the file until the first time GetToken or GetTokenAsync is
called. In cases where there is a problem with decoding the file, we
ensure `CredentialUnavailableException` is raised (this ensures things
like `ChainedTokenCredential` and `AzureDefaultCredential` are able to
correctly ignore these credentials, if needed.

Contributes To: #8309

azure-sdk committed:
Sync eng/common directory with azure-sdk-tools repository (#10240)

kinelski committed:
[EventHubs] Distributed Tracing: Stamp EventHubs info on processing and message spans (Track Two) (#10226)

* Added endpoint and EH info to Message scopes

* Added endpoint and EH info to Processing scopes

* Fixed event sources with wrong enumeration

kinelski committed:
Included enqueued time attribute to processing scope (#10227)

JoshLove-msft committed:
Api Updates (#10232)

tg-msft committed:
Mark Search tests as inconclusive until we add tests.yml (#10243)

It looks like the Search tests are getting pulled into the core - ci pipeline so
we'll just mark them as Inconclusive if not configured to run live.

pakrym committed:
Add readmes and changelogs to extensions. (#10215)

matthchr committed:
Include Batch management tests when building (#10246)

JoshLove-msft committed:
Implement receive deferred messages (#10235)

xaliciayang committed:
Adding LTR for MI to .NET SDK (#10129)

* adding LTR MI tests for net

* adding test session records

* adding generated files

* adding another generated file

* fixing things that should not have been updated

* additional fix; for API that is not ready

* making sure versions are correct in SdkInfo

* add updated test recordings; add skip tags to some tests

* fix a version

* session records

* reset serverazuread version

* add databasestate files

* make tests and records consistent

* other required files

* rearranging

* Update sdk/sqlmanagement/Microsoft.Azure.Management.SqlManagement/src/Microsoft.Azure.Management.Sql.csproj

* bump version

Co-authored-by: Yeming Liu <felix_liu@outlook.com>

christothes committed:
Fix flaky tests and discovered bugs for EventProcessorClient (#10182)

* fix tests and discovered bugs

* whitespace

* introduce EventHubsExceptionExtensions to get/set extended exception data

* spelling fix

* refactor EventHubsExceptionExtensions and revert changes to EventHubsException

* partitionsBeingProcessed >= partitionIds.Length

ziwa-msft committed:
Add MinimalTlsVersion Property to MI and Sql DB APIs (#10179)

* regenerated sql sdk based on latest swagger

* Add MinimalTlsVersion Property in Server and Managed Instance

* Add tests for minimalTlsVersion property of a Logical Server and Managed Server

* fixing missing ;

* fix bug in DB test

* TRying to fix the build break with manageddatabaserestoredetails

* Revert "TRying to fix the build break with manageddatabaserestoredetails"

This reverts commit 6d8b73b7e1155ee1cf4439fd84825a220b947a2d.

* Regen SDK after Build Break Fix

* Updated ServerTest and Recordings

* fix test

* metadata change

* fixing bad merge

* removing unrelated files

* removing unrelated files

* removing unrelated files

* remove unrelated change

* fix bad merge

* removing unrelated files

* remove unrelated files

* remove unrelated files

* Update version and release notes

* Move sql_resource-manager and update AssemblyInfo

* update sql_resource-manager.txt

Co-authored-by: Andrija Cicovic <60783126+cicovica@users.noreply.github.com>

seanmcc-msft committed:
Fixed issue where DataLakePathClient.Create() and CreateIfNotExists() were not public (#10257)

ellismg committed:
[EventHubs] Clean up EventPosition.Equals logic (#10252)

Our logic for implementing `EventPosititon.Equals` had some extra code
that would try to use `object.ReferenceEquals` as a way to see if one
EventPosition was equal to another.

However, `EventPosition` is a struct, so it's not clear how this check
could ever really return true, since our calls to `ReferenceEquals` is
going to box both `EventPosition` arguments, so they will never be in
the same place in memory.

andrueastman committed:
[Identity] [Fix] UsernamePasswordCredential call with correct parameter order (#10187)

* Call UsernamePassword with correct parameter order

seanmcc-msft committed:
Made DataLake client pipelines internal (#10271)

jsquire committed:
[Event Hubs Client] `Event Processor<TPartition>` Skeleton (#10229)

The focus of these changes is to flesh out the public skeleton of the
`EventProcessor<TPartition>` to match the design proposal, and author the
public-facing doc comments to accompany them.

After these changes, the full set of types for the event processor should
exist and be in the correct project/namespace to allow for implementation of
the behaviors to move forward.

chidozieononiwu committed:
Prep for Release of Microsoft.Azure.ServiceBus (#10277)

chidozieononiwu committed:
Update Microsoft.Azure.ServiceBus to version 4.1.2 (#10279)

heaths committed:
Update versions for servicing release (#10283)

* Update versions for servicing release

* Fix previous releases' dates in CHANGELOG

kinelski committed:
[EventHubs] Distributed Tracing: Linked send scopes to message scopes (Track Two) (#10267)

serkantkaraca committed:
Rolling back to WindowsAzure.Storage due to timeout issues. (#10265)

angosms committed:
Regenerate Managed Identity SDK from updated swagger (#9915)

* Regenerate SDK from updated swagger. Fix tests.

* Run generate.ps1

* Add new files

maririos committed:
review docstrings and readme (#10239)

maririos committed:
Improve performance by pre-encoding JSON property names in serializer (#10284)

* Improve performance by pre-encoding JSON property names in serializer

* Code review fixes

maririos committed:
Renaming offset and length to with prefix grapheme (#10238)

* renaming

* update docstrings

danieljurek committed:
Smoke Tests: add logic for using latest package versions (#10214)

* add logic for using latest package versions
* clean up yml
* add restore step
* try with dotnet restore wildcards and optional commands
* use better Update-Dependencies strategy, add Test-DependencyDevVersions validation
* include dotnet restore
* modify scripts, use pwsh
* use the latest Azure.Core
* add Azure.Messaging.EventHubs.Processor
* update event hubs smoke test
* add $PACKAGE_REFERENCE_XPATH to Test-DevDependencyDevVersions
* pin a verison for Microsoft.Azure.DocumentDB.Core
* update version specifier for Azure.Messaging.EventHubs.Processor, extra commentary in Update-Dependencies.ps1 describing why all packages have to be listed from a sleet source
* collapse dev version validation into Update-Dependencies.ps1, simplify pipeline yaml
* output warnings in Azure DevOps based on -CI flag
* use -CI flag

chidozieononiwu committed:
Increment package version after release of Microsoft.Azure.ServiceBus (#10281)

JoshLove-msft committed:
More API updates (#10292)

kinelski committed:
[EventHubs] Apply Retry Policy to CreateBatchAsync (Track Two) (#10320)

jsquire committed:
[Event Hub Client] Event Processor<T> Initialization (#10317)

The focus of these changes is to begin building out the `EventProcessor<T>`
implementation, starting with private infrastructure and initialiazing state
during construction.

weshaggard committed:
Sync eng/common directory with azure-sdk-tools repository (#10323)

matthchr committed:
Update Batch DataPlane SDK to support new REST API version (#10290)

* Regenerate protocol from latest Swagger

* Update Batch SDK Object Model

* Update tests

* Add new error code strings

* Update Certificate creation factory methods

  - Make it more clear that password is optional for PFX formatted
    certificates.

* Update Batch changelog

maririos committed:
Arch board feedback (#10308)

* arch board feedback

* add missing  change to changelog

* remove rounding of scores

* add new feature to changelog

* PR feedback

* remove label from docstrings

* extra enter

ShivangiReja committed:
[Service Bus] Add message batching (#10305)

[Service Bus] Add message batching

pakrym committed:
Update the code generator version (#10303)

pakrym committed:
Azure.Core 1.0.3 changelog (#10336)

annelo-msft committed:
Set up new Unified Pipeline for Form Recognizer (#10345)

* add ci.yml for form recognizer

* updating artifacts

pakrym committed:
Add code generation to PR code checks (#10315)

azure-sdk committed:
Increment version for extensions releases (#10247)

* Increment package version after release of Azure.AspNetCore.DataProtection.Blobs

* Increment package version after release of Azure.Extensions.Configuration.Secrets

* Increment package version after release of Azure.AspNetCore.DataProtection.Keys

maririos committed:
[Text-analytics] Renames and more (#10341)

* renames and more

* modify package version

* update date of release in changelog

azure-sdk committed:
Increment package version after release of Microsoft.Azure.Batch (#10348)

pakrym committed:
Implement RequestFailedException serialization. (#10335)

weshaggard committed:
Fix spacing char in ci.yml (#10362)

replace utf-8 char 160 (non-breaking-space) with normal utf-8 32 space

chidozieononiwu committed:
Enable ReleaseNotes Extraction for Track 1 Libraries (#10301)

pakrym committed:
Update Azure.Core version to 1.1.0 (#10367)

kinelski committed:
[Azure.Core Shared] Adding link to Diagnostic Scope should take attributes (#10368)

pakrym committed:
Add @pakrym to extensions codeowners (#10356)

pakrym committed:
Only run code checks related to the service directory (#10357)

azure-sdk committed:
Increment version for core releases (#10371)

pakrym committed:
Update Azure.Core to 1.1.0 (#10372)

areddish committed:
Custom Vision Client SDK updates (#9904)

* Add Custom Vision 3.2

* Test updates

* Update session records
Update ids.txt

* Add release notes, bump the version

kinelski committed:
[EventHubs, Azure.Core Shared] Move enqueuedTime from root activity to linked activity (Track Two) (#10374)

* Fix enqueuedTime behavior

* Removed 'Ignore' attribute

* Addressed comments

kinelski committed:
[EventHubs] Partition Receiver Options (Track Two) (#10338)

* Added implementation to Partition Receiver Options

* Added Partition Receiver Options Tests

AlexanderSher committed:
Add sku parameter to test-resource.json (#10381)

heaths committed:
Add 7.1-preview service features for Key Vault 4.1-preview.1 (#10225)

* Prepare Key Vault version 4.1-preview

* Add "import" option for key_ops

Resolves #10063

* Add RecoverableDays property for soft delete-enabled vaults

Resolves #10046

* Update CHANGELOGs

* Update API documentation

* Resolve PR feedback

* Update APIs

* Update tests for 7.1-preview

christothes committed:
StopProcessingAsyncStopsProcessingForEveryPartition (#10397)

AlexanderSher committed:
Update Microsoft.AppConfiguration/configurationStores API version  (#10398)

* Add sku parameter to test-resource.json

* Update API version

rileymckenna committed:
Made microsoft session address non-readonly (#10199)

* fixed unit tests reran the generate.ps1

* Delete code-model-v1

Co-authored-by: Yeming Liu <Yeming.Liu@microsoft.com>

laurawu19 committed:
Upgrading diskRP .net sdk to use version 2019-11-01. The Microsoft Az… (#10217)

* Upgrading diskRP .net sdk to use version 2019-11-01. The Microsoft Azure Compute Management library will now be at version 33.0.0

* rerecording tests

* fixing Disk_CRUD_ImportDisk

* Rerecord tests

* Update api tags in AzSdk.RP.props

Co-authored-by: REDMOND\jopolloc <jopolloc@microsoft.com>

pakrym committed:
Update Autorest C# (#10384)

pakrym committed:
Remove test workaround (#10427)

serkantkaraca committed:
Don't contrain storage I/O at timeout for checkpointng calls. (#10429)

heaths committed:
Fix Key Vault project files (#10395)

* Fix Key Vault project files

* Remove test strategy breaking build

The Windows_NetCoreApp_ProjectReferences test strategy is breaking the
build because it's converting PackageReferences to ProjectReferences,
and for recorded tests this is failing the build beacuse Secrets is now
a preview version, which then causes Azure.Identity to fail because
records are wrong.

* Hard code SecretClient to use service version 7.0

Otherwise, converting PackageReferences to ProjectReferences causes recorded tests to fail using the latest SecretClient version from a preview we just shipped.

This also reverts commit c9fa0a01363ffc55eef39e3d6c3430bb8e888dd7.

maririos committed:
Add missing S to SentimentConfidenceScore (#10433)

* add missing S

* changelog

ShivangiReja committed:
[Service Bus] Add documentation for all methods related to receiving (#10418)

[Service Bus] Add documentation for all methods related to receiving

heaths committed:
Enable live tests for Search (#10375)

* Enable live tests for Search

* Add required "resources" section

* Optionally exclude track 1 projects

* Remove data plane tests from nightly live runs

Commit:
Azure.Identity updating DefaultAzureCredential to enable authenticating through the Azure CLI (#10380)

* Add CliCredential

* Update newsoft version

* Update

* Update CliCredential.cs

* Update CLICredential.cs

* Update CLICredential.cs

* Update CLICredential.cs

* Update CLICredential.cs

* Update CliCrdential.cs

* Update CliCredential.cs

* Update CLICredential.cs

* Update CLICredential.cs

* Update CliCredential.cs

* Update CliCredential.cs

* Add Nunit Test of CliCredential

* Revert "Add Nunit Test of CliCredential"

This reverts commit 20603bdba4939980934c52d45c9d7aeb3dffa17d.

* Add UT of CliCredential

* Update UT

* Update UT

* Update UT

* Update UT

* Update UT

* Wrap Process with using, ParseExact ExperisOn

* Add validation of ScopeToResource, Add "--ouput json" para in command.

* Update validation error message filed

* Rename CliCredential to AzureCliCredential, and some vairables name changed.

* [WIP] adding DeveloperCredential

* refactoring AzureCliCredentialClient

* fixing tests

* restricting path for AzureCliCrednential

* updating AzureCliCredential to only execute az from PATH

* removing developer credential prototype code

* Fixing async linter errors

* Updating API files

* updating changelog

* addressing PR feedback

* removing AccessTokenUtilities

* addressing more PR feedback

* addressing overlooked Regex comment

* Update to execute shell out of system dir

Co-authored-by: wantedfast <inaccuarcy@gmail.com>
Co-authored-by: Geralt Wang <v-xinlwa@microsoft.com>

weshaggard committed:
Remove merge=union for csproj files (#10440)

Recently seen some bad merges because of merge=union and cannot figure out any reach why we would need it so remove it. Setting the csproj and sln file gitattributes to match that of the .NET team https://github.com/dotnet/runtime/blob/master/.gitattributes#L58

annelo-msft committed:
FormRecognizer preview 1 prototype for UX study (#10319)

* intial generation

* first authenticated call

* start training LRO is minimally functional

* map Model to CustomModel

* interim commit

* start extract form LRO is minimally functional

* map AnalyzeResult.PageResult to ExtractedForm

* start train form with labels minimally functional

* map Model to CustomLabeledModel

* map DocumentResult to ExtractedForm

* tidy up of public API

* setup for ReceiptClient; adding async variants of LROs to CustomFormClient

* Extract Receipt functioning

* Map AnalyzeResult to ExtractedReceipt

* Functional extract layout

* Map AnalyzeResult to ExtractedTable

* Add Tables to ExtractedForm

* Make internal types internal again

* some renames

* Added Uri overloads

* tweak

* first success with OCR/references

* adding OCR values and mapping types as workarounds

* move custom types from Models to Custom namespace

* interim checkin

* Completed first iteration of bounding box on ExtractedReceipt. Removed AsReadOnly() calls for IReadOnlyList

* updates per APIView feedback

* bug fixes

* initial prototype code for FR

* update solution file

* intermediate checkin for TODOs

* interim checkin for TODOs

* revert template files

* cleaning up /template files

* Updating TODOs with links to github issues

* samples project change

* rebase to new CodeGen

* tidy up

* git attributes for gen code

* pr feedback

* bug fix

* bug fix

* bug fix

* stop tracking samples

* pr feedback

* add api

* remove sln under sdk/cognitiveservices

ellismg committed:
[Identity] Add fix from #10187 to CHANGELOG (#10274)

pakrym committed:
Update Autorest C# (#10442)

heaths committed:
Update Key Vault 4.1.0-preview.1 release dates (#10474)

azure-sdk committed:
Increment version for keyvault releases (#10475)

* Increment package version after release of Azure.Security.KeyVault.Keys

* Increment package version after release of Azure.Security.KeyVault.Secrets

* Increment package version after release of Azure.Security.KeyVault.Certificates

amchandn committed:
Removing deprecated BackupVault related sdk tests (#10423)

* Removing GetAndUpdateBackupVaultAndStorageConfigTest

* Revert "Removing GetAndUpdateBackupVaultAndStorageConfigTest"

This reverts commit b0784a144142032ca823e417274c36fcd2503b32.

* Removing BackupVault related code

aim-for-better committed:
1. Generate sdk with azure-arm:true (#10424)

Co-authored-by: Zhenyu Zhou <zhezhou@microsoft.com>

jsquire committed:
[Event Hubs Client] Event Processor<T> Startup/Shutdown (#10403)

The focus of these changes is to implement the startup and shutdown functionality
for the event processor, along with enhancing the tests to cover the previously
implemented constructors.

At this time, partition processing and load balancing are not implemented;
these areas remain stubbed to allow for testing and will be part of a new
workstream.

annelo-msft committed:
Addressing APIView Feedback (#10482)

* issues 10468, 10467

* 10466, 10462, 10458, 10406

* fixes

* pr fb

* pr feedback

* fixes

* fixes

* update api listing

Commit:
Azure.Identity adding KnownAuthorityHosts constants (#10280)

* Azure.Identity adding KnownAuthorityHosts constants

* updating api listing

* addressing PR feedback

* fixing US casing

* updating API listing

Commit:
Azure.Identity update changelog for 1.2.0-preview.1 release (#10500)

* Azure.Identity update changelog for 1.2.0-preview.1 release

* added KnownAuthorityHosts rel note

Commit:
Azure.Identity Fixing AzureCliCredential check for no logged in account (#10497)

* Fixing AzureCliCredential check for no logged in account

* switch to IndexOf instead of Contains to specify StringComparison type

azure-sdk committed:
Increment package version after release of Azure.AI.TextAnalytics (#10491)

seanmcc-msft committed:
Attempting to fix Storage CI (#10502)

Commit:
updating identity CODEOWNERS entry to include all subfolders (#10514)

jongio committed:
Change accquring to acquiring (#8960)

jongio committed:
Fix typo in probing message (#8947)

pakrym committed:
Allow flushing read-only streams produced by SDK (#10508)

azure-sdk committed:
Increment package version after release of Azure.Identity (#10505)

heaths committed:
Exclude some attributes from API listing (#10515)

* Exclude some attributes from API listing

The CompilerGeneratedAttribute and EditorBrowsableAttribute are noisy in the API listing and not really necessary.

* Keep EditorBrowsableAttribute

...based on PR feedback

annelo-msft committed:
Address Form Recognizer APIView Feedback (#10504)

* 10459, 10492

* 10939

* 10463

* 10540

* update api

* update api

kinelski committed:
[EventHubs] Partition Receiver: constructors and CloseAsync (Track Two) (#10494)

seanmcc-msft committed:
Made FileSystemClient.GetRootDirectory() internal, improved PathClien… (#10546)

seanmcc-msft committed:
Prepared Storage for March 2020 release (#10436)

kinelski committed:
[EventHubs] Partition Receiver: GetPartitionPropertiesAsync (Track Two) (#10548)

JoshLove-msft committed:
Exceptions and Event Source Logging (#10460)

wantedfast committed:
Keep git-fork update

wantedfast commented:
Since AzureCliCredentila merged already, colse this one
